src/client/common/installer/productInstaller.ts

```diff
@@ -1,34 +1,74 @@
 /* eslint-disable max-classes-per-file */
 
 import { inject, injectable, named } from 'inversify';
-import { CancellationToken, OutputChannel, Uri } from 'vscode';
+import { CancellationToken, Memento, OutputChannel, Uri } from 'vscode';
 import { IPythonInstaller } from '../../api/types';
 import '../../common/extensions';
 import * as localize from '../../common/utils/localize';
 import { Telemetry } from '../../datascience/constants';
+import { InterpreterPackages } from '../../datascience/telemetry/interpreterPackages';
 import { IInterpreterService } from '../../interpreter/contracts';
 import { IServiceContainer } from '../../ioc/types';
 import { PythonEnvironment } from '../../pythonEnvironments/info';
+import { getInterpreterHash } from '../../pythonEnvironments/info/interpreter';
 import { sendTelemetryEvent } from '../../telemetry';
 import { IApplicationShell, IWorkspaceService } from '../application/types';
 import { STANDARD_OUTPUT_CHANNEL } from '../constants';
+import { traceError } from '../logger';
 import { IProcessServiceFactory, IPythonExecutionFactory } from '../process/types';
 import {
+    GLOBAL_MEMENTO,
     IConfigurationService,
     IInstaller,
+    IMemento,
     InstallerResponse,
     IOutputChannel,
     IsCodeSpace,
     ModuleNamePurpose,
     Product
 } from '../types';
+import { sleep } from '../utils/async';
 import { isResource } from '../utils/misc';
 import { StopWatch } from '../utils/stopWatch';
 import { ProductNames } from './productNames';
 import { InterpreterUri, IProductPathService } from './types';
 
 export { Product } from '../types';
 
+export async function trackPackageInstalledIntoInterpreter(
+    memento: Memento,
+    product: Product,
+    interpreter: InterpreterUri
+) {
+    if (isResource(interpreter)) {
+        return;
+    }
+    const key = `${getInterpreterHash(interpreter)}#${ProductNames.get(product)}`;
+    await memento.update(key, true);
+}
+export async function isModulePresentInEnvironment(memento: Memento, product: Product, interpreter?: InterpreterUri) {
+    if (isResource(interpreter)) {
+        return;
+    }
+    const key = `${getInterpreterHash(interpreter)}#${ProductNames.get(product)}`;
+    if (memento.get(key, false)) {
+        return true;
+    }
+    const packageName = translateProductToModule(product);
+    const packageVersionPromise = InterpreterPackages.getPackageVersion(interpreter, packageName)
+        .then((version) => (typeof version === 'string' ? 'found' : 'notfound'))
+        .catch((ex) => traceError('Failed to get interpreter package version', ex));
+    try {
+        // Dont wait for too long we don't want to delay installation prompt.
+        const version = await Promise.race([sleep(500), packageVersionPromise]);
+        if (typeof version === 'string') {
+            return version === 'found';
```

Comment:
Do you want to add it to the memento here since we found it?

Reply:
Wasn't sure if it was a good idea.
What if user uninst, re-created the environment, and testing for us also gets more messy.